### DISCUSSION TITLE
Testing escaping shadow interposition

### DISCUSSION THREAD

**Discussion by akumar1512** [shadow/shadow#2685] [Q&A] [answered]
I wanted to test the functionality of shadow switching interposition from an application. While reading #2157 , I wanted to try from one of the way @sporksmith suggested [here](https://github.com/shadow/shadow/issues/2157#issuecomment-1242820198).

I have added two custom syscalls - `SYS_allow_native` and `SYS_deny_native` in `src/main/host/syscall_number.h` and using them in my test application in the following way -

`testapp.c`

```
#include <stdio.h>
#include <unistd.h>
#include <sys/syscall.h>

#define SYS_allow_native 1006
#define SYS_deny_native 1007

int main(void) {

        printf("Turning interpose OFF!\n");

        syscall(SYS_allow_native); // SYS_allow_native syscall

        printf("Testing sleep for two seconds\n");
        sleep(2);
        printf("Turning interpose ON!\n");

        syscall(SYS_deny_native); // SYS_deny_native syscall

        printf("Testing sleep for two seconds after interpose turned ON\n");
        sleep(2);
        printf("Test completed\n");
        return 0;
}
```

I added cases to handle these syscalls in `src/main/host/syscall_handler.c` in `SysCallReturn syscallhandler_make_syscall(SysCallHandler*, const SysCallArgs*)` inside switch.

I wanted to call `shim_swapAllowNativeSyscalls` inside the cases but when I add `#include "lib/shim/shim.h"` to `src/main/host/syscall_handler.c` the build fails with undefined reference/linker error.

I read in the discussion that we can only call functions in the "three" preloaded libraries - `injector, libc, openssl`

Is there a way I can still call this API through `injector`. Any help in suggesting the way in which such syscall should be implemented would be great help!

Thanks

**Comment by stevenengler** [comment, marked as answer]
In Shadow there are two places you can intercept and handle syscalls: in the main shadow process (in `src/main/`), and in the shim (in `src/lib/shim/`). The shim is the library that's preloaded into each managed process. The `shim_swapAllowNativeSyscalls` function can only be called from the shim library and not from the main shadow process.

You'll want to intercept these new syscalls in the shim and not in the main shadow process. So instead of adding handlers in `src/main/host/syscall_handler.c`, you should probably handle the syscalls in `shim_syscallv` in `lib/shim/shim_syscall.c`.

**Reply by akumar1512** [reply]
Thanks @stevenengler , I have tried this. This works. 
However seems like I can also access the symbol's address using [dlsym](https://man7.org/linux/man-pages/man3/dlsym.3.html), since I have it already in the process' address space. I am calling it using function pointer directly from plugin.

**Reply by stevenengler** [reply]
That's good to know that method works too. Are you just calling `dlopen` on the shim "libshadow_injector.so" path, and then calling `dlsym` to get the address of `shim_swapAllowNativeSyscalls`?

**Reply by stevenengler** [reply]
Also is there a particular reason you need to disable interposition? Is there some feature that you need that Shadow's missing?

**Reply by hammer498** [reply]
Re: our goal for disabling interposition, it seemed easier to start a new discussion topic, done [here](https://github.com/shadow/shadow/discussions/2689) since we are also curious if you have any feedback.

**Reply by akumar1512** [reply]
And regarding `dlopen`, we don't need it since `libshadow_injector.so` is already pre-loaded and helps link the shared lib during initialization into process' address space.

**Comment by akumar1512** [comment]
A short snippet using `dlsym`. You don't need to re-build shadow if disabling interpose using this method.

```
#include <dlfcn.h>
#include <stdbool.h>
...
void *dlsym(void *handle, const char *symbol);
typedef bool(*disable_ip)(bool);

...
int main(void){

    disable_ip ip = dlsym(NULL, "shim_swapAllowNativeSyscalls");
    ...

    ip(true);
    // syscall to real OS
    ...
    ip(false);
    ...

    return 0;
}
```
